Standardized API Service Access

Allcharles opened the pull request:
Abstracted the logic from model services (sites, projects, user) into one model service file. Sites/Projects/User services now extend that file which allows for more in-depth testing and far better support for filtered requests.

Things to review

- `model.service.ts` api logic
  - Check correct HTTP request method (GET/POST/PATCH)
  - Check filter logic on details and show functions are correct
- `model.service.spec.ts` unit tests
- `sites.service.ts` api logic
  - Do the functions use the correct model service function?
- `projects.service.ts` api logic
  - Do the functions use the correct model service function?
- `user.service.ts` api logic
  - Do the functions use the correct model service function?
- baw-api/mock
  - Check mock services have the same public functions as their originals

Closes: #65 
Closes: #64

Allcharles committed:
Abstracted api services to model service

Allcharles committed:
Working unit tests

Allcharles committed:
Added filter to api routes

Allcharles committed:
Added list of model service tests

Allcharles committed:
Added model service details unit tests

Allcharles committed:
Added model service show unit tests

Allcharles committed:
Added model service new unit tests

Allcharles committed:
Added missing output error handling unit tests

Allcharles committed:
Updated project service unit tests

Allcharles committed:
Updated site service unit tests

Allcharles committed:
Updated user service unit tests

Allcharles committed:
Updated Mocks

Allcharles commented on src/app/models/Project.ts:
```diff
@@ -86,3 +86,11 @@ export class Project implements ProjectInterface {
   }
 }
+
+export const mockProject = new Project({
+  id: 1,
+  name: "name",
+  description: "description",
+  creatorId: 1,
+  siteIds: new Set([1])
+});
```
Creating a mock project may be a bad practice for the unit tests which implement it

atruskie replied:
Generally faking frameworks can help us generate better test data: https://github.com/marak/faker.js

Allcharles commented on src/app/models/Site.ts:
```diff
@@ -82,3 +82,12 @@ export class Site implements SiteInterface {
   }
 }
+
+export const mockSite = new Site({
+  id: 1,
+  name: "name",
+  description: "description",
+  creatorId: 1,
+  projectIds: new Set([0]),
+  imageUrl: "/assets/images/site/site_span4.png"
+});
```
Creating a mock site may be a bad practice for the unit tests which implement it

Allcharles commented on src/app/services/baw-api/mock/baseApiMockService.ts:
```diff
@@ -1,12 +1,17 @@
-  public getSessionUser() {}
+  public getSessionUser(): SessionUser | null {
+    return null;
+  }
```
Don't need function return typing in mock services

Allcharles commented on src/app/services/baw-api/model.service.ts:
```diff
@@ -0,0 +1,143 @@
+  }
+}
+
+export type Args = (string | number)[];
```
Missing type XML comment

atruskie commented on src/app/services/baw-api/model.service.spec.ts:
```diff
@@ -0,0 +1,2565 @@
+      (err: APIErrorDetails) => {
+        expect(err).toBeTruthy();
+        expect(err).toEqual({
+          stat
```
I've managed to understand a little of this file, but:

- it is way too long
- has way too much repetition in it

Can you find some ways to break it up?

Allcharles replied:
I have two immediate methods I could use to split up the unit tests. Either I create multiple `describe` functions inside the file, or I can split the file into multiple smaller files.

The first method will maintain a large file, however I can group unit tests by the method it tests, and the unit test output will separate the tests into different groups similar to separate unit test files. 

The second method of splitting the file up would have a similar effect to the first method in terms of output from the unit testing framework, however it would add unnecessary files to the directory. I would likely split the files by method name resulting in the following file name format: `model.<method>.spec.ts`.

Which method do you prefer.

As for the repetition, some of the no filter vs filter tests could be removed with regards to error handling (as that is independent of filter), however this would make the assumption that someone wont break that independence in the future which I'm not a huge fan of. While I understand there is a lot of repetition, I'm unsure how I would reduce it without reducing the coverage of the testing.

atruskie replied:
I'll let you decide.

If you can factor out enough common code, then keep it one file.

Normally I'd baulk at more files, but 2.5K lines in a file are unwieldy. Breaking out into methods is fine if you can't factor out helper methods/common code etc

atruskie commented on src/app/services/baw-api/model.service.ts:
```diff
@@ -0,0 +1,143 @@
+import { APIErrorDetails } from "./api.interceptor";
+import { BawApiService, Filters } from "./base-api.service";
+
+export let MOCK_CLASS_BUILDER = new InjectionToken("test.class.builder");
```
This seems like a conflation of concerns. Why is a test token defined in a non-test file? Is there a better spot for it?

Allcharles replied:
The injection token is used by the model.service.spec.ts file so moving this variable outside of the model.service file would be a bit awkward. Potentially i can move it to the injection token definition to the unit testing file and import it into the service file?

atruskie replied:
I think my biggest issue is one of naming then. "Mocks" shouldn't exist in non-test code.

atruskie commented on src/app/services/baw-api/model.service.ts:
```diff
@@ -0,0 +1,143 @@
+  constructor(
+    http: HttpClient,
+    config: AppConfigService,
+    @Inject(MOCK_CLASS_BUILDER) private classBuilder: (object: any) => T
```
Why does this need to be injected? Why is it a 'mock' builder?

Allcharles replied:
`MOCK_CLASS_BUILDER` is used by the model.service unit tests. The reason it is a mock is because in the actual application `ModelService` is never actually used, only extended from. Any non service inputs in a service must have an injector so that the app.module file can instantiate default values. Thus to test the file, i need a mock injector which is only ever used by the model service unit tests.

atruskie replied:
See [comment](https://github.com/QutEcoacoustics/workbench-client/pull/67#discussion_r368373182) on naming. What about "stub" or "abstract"?

atruskie commented on src/app/services/baw-api/model.service.ts:
```diff
@@ -0,0 +1,143 @@
+   * @param path URL path
+   * @param args Arguments to insert
+   */
+  private refineUrl(path: string, args: Args): string {
```
Is it possible to reuse angular's URL templating functions? We should do so

Allcharles replied:
As far as I can tell, there is no exposed method of handling URL attributes. HttpClient only exposes a method of adding HTTPParams which are the attributes placed after the `?` in a url. Eg. `https://brokenlink?http_params_go_here`.

atruskie replied:
https://angular.io/api/router/Router#parseurl

atruskie commented on src/app/services/baw-api/projects.service.ts:
```diff
@@ -9,71 +9,49 @@ import {
+
   constructor(http: HttpClient, config: AppConfigService) {
-    super(http, config);
+    const classBuilder = (project: ProjectInterface) => new Project(project);
```
is this needed for every model?

Let's make it optional. Doing a `new T()` should be possible inside of `ModelService`

Allcharles replied:
Seems to be possible, although not as clean as you are thinking: https://stackoverflow.com/questions/17382143/create-a-new-object-from-type-parameter-in-generic-class

atruskie replied:
Yep, you're right.

I'd suggest following this pattern:

```typescript
//model
class foo {
    constructor() {
    }

    name: string;
}

abstract class CommonApi<T> {
    abstract modelFactory(): T;
    

    get() {
        return this.modelFactory();
    }
}

class FooService extends CommonApi<foo> {
    modelFactory() { return new foo() };

}
```

Allcharles replied:
I found what I consider to be a cleaner solution in this commit: https://github.com/QutEcoacoustics/workbench-client/pull/67/commits/fe59ca29ed64861a5096f2aa1226580fbf0a1341.

Basically it allows me to extend the `ApiCommon` class like this:

```javascript
class ApiCommon<T> {
  constructor(
    http: HttpClient,
    config: AppConfigService,
    private type: new (object: any) => T
  ) {
    super(http, config);
  }
}

class ProjectsService extends ApiCommon<Project> {
  constructor(http: HttpClient, config: AppConfigService) {
    super(http, config, Project);
  }
}
```

atruskie reviewed (changes requested):
There seems like a lot of repetition in the `xxxMockService` classes. I think this can be simplified but it's really a smell of another issue.

I like the common functionality abstraction. However, why don't you use common names as well? Why `getProjects`, `getProject`, etc over just `get` and `getAll`? Oh I see why you're doing this. I'd suggest trying overriding methods via inheritance rather than doing a bunch of plumming in every child class.

We use the common names `list` (GET /x), `show` (GET /x/123), `destroy`, `create`, and `update` in rails - it seems like it would be a good idea to reuse those names. I'd say that using "details" as the name for "list" is confusing.

This has all been done before in `$resource` in Angular 1.x. I'd suggest reading the docs https://docs.angularjs.org/api/ngResource/service/$resource for some history.

Your idea of abstracting service commonality into one place is great but needs some work.

- I'd avoid the name `ModelService` - its mixing two very general names and is not clear by itself. What about `WorkbenchApiService`, or `ApiCommon`, or `ApiService`?
- I'd customize each child class with route parameters when they're instantiated and using that customization for storing path and host data, rather than passing the path/URL into each service method as a parameter

Allcharles committed:
Refactored details functions to list in api

Refactored apiDetails and details functions in Model Service and BawApi Service classes to use 'list' keyword instead of 'details.

Allcharles committed:
Renamed ModelService to ApiCommon

Allcharles committed:
Rename MOCK_CLASS_BUILDER

Relates to review comment: https://github.com/QutEcoacoustics/workbench-client/pull/67#discussion_r368373182

Allcharles committed:
Generic class builder

Related to review comment: https://github.com/QutEcoacoustics/workbench-client/pull/67#discussion_r368360804

Allcharles committed:
Updated unit tests to accept generic class builder

Allcharles committed:
Split out api-common unit test files

Split out api-common unit tests into multiple smaller files to better readability.

Relates to review comment: https://github.com/QutEcoacoustics/workbench-client/pull/67#discussion_r368358914

Allcharles committed:
Fixed @allcharles reviews

Fixed the following comments:

- https://github.com/QutEcoacoustics/workbench-client/pull/67#discussion_r364541700
- https://github.com/QutEcoacoustics/workbench-client/pull/67#discussion_r364542231

Allcharles committed:
Improved api-common refinedUrl function

Replaced api-common refineUrl function body with router method for extracting url arguments.
Relates to review comment: https://github.com/QutEcoacoustics/workbench-client/pull/67#discussion_r368360494
